pydantic/_internal/_generate_schema.py

```diff
@@ -0,0 +1,592 @@
+            required = field_name in required_keys
+
+            if get_origin(annotation) == _typing_extra.Required:
+                required = True
+                annotation = get_args(annotation)[0]
+            elif get_origin(annotation) == _typing_extra.NotRequired:
+                required = False
+                annotation = get_args(annotation)[0]
+
+            field_info = FieldInfo.from_annotation(annotation)
+            fields[field_name] = self.generate_field_schema(
+                field_name, field_info, validation_functions, required=required
+            )
+
+        return core_schema.typed_dict_schema(fields, extra_behavior='forbid')
+
+    def _namedtuple_schema(self, namedtuple_cls: Any) -> core_schema.CallSchema:
+        """
+        Generate schema for a NamedTuple.
+        """
+        annotations: dict[str, Any] = _typing_extra.get_type_hints(namedtuple_cls, include_extras=True)
+        if not annotations:
+            # annotations is empty, happens if namedtuple_cls defined via collections.namedtuple(...)
+            annotations = {k: Any for k in namedtuple_cls._fields}
+
+        arguments_schema = core_schema.ArgumentsSchema(
+            type='arguments',
+            arguments_schema=[
+                self._generate_parameter_schema(field_name, annotation)
+                for field_name, annotation in annotations.items()
+            ],
+        )
+        return core_schema.call_schema(arguments_schema, namedtuple_cls)
+
+    def _generate_parameter_schema(
+        self,
+        name: str,
+        annotation: type[Any],
+        mode: Literal['positional_only', 'positional_or_keyword', 'keyword_only'] | None = None,
+    ) -> core_schema.ArgumentsParameter:
+        """
+        Prepare a ArgumentsParameter to represent a field in a namedtuple, dataclass or function signature.
+        """
+        field = FieldInfo.from_annotation(annotation)
+        assert field.annotation is not None, 'field.annotation should not be None when generating a schema'
+        schema = self.generate_schema(field.annotation)
+        schema = apply_metadata(schema, field.metadata)
+
+        parameter_schema = core_schema.arguments_parameter(name, schema)
+        if mode is not None:
+            parameter_schema['mode'] = mode
+        if field.alias is not None:
+            parameter_schema['alias'] = field.alias
+        return parameter_schema
+
+    def _generic_collection_schema(self, type_: Any) -> core_schema.CoreSchema:
+        """
+        Generate schema for List, Set etc. - where the schema includes `items_schema`
+
+        e.g. `list[int]`.
+        """
+        try:
+            name = type_.__name__
+        except AttributeError:
+            name = get_origin(type_).__name__  # type: ignore[union-attr]
+
+        return {  # type: ignore[misc,return-value]
+            'type': name.lower(),
+            'items_schema': self.generate_schema(get_first_arg(type_)),
```

Comment:
Absolutely, but even better, constraints that don't apply to a list should be automatically forwarded to it's items, so

```py
data: Annotated[list[int], Field(min_items=3, gt=0)]
# or
data: list[int] = Field(min_items=3, gt=0)
```
Should both mean the same as your example.

Reply:
Awesome!

In that case I wonder what would happen some layers down, e.g.

```Python
data: Annotated[tuple[int, list[int], list[Annotated[float, Lt(3)]]]]], Field(min_items=3, gt=0)]
```

would `min_items` apply to the inner list? And would `gt` apply to the inner `int`? And would the top level `gt` in `Field()` conflict with the inner `float`'s `lt` in any way?

Reply:
The code (so far) is

https://github.com/pydantic/pydantic/blob/594effa279668bd955e98f1cd5c036b37d3bbd40/pydantic/_internal/_generate_schema.py#L558-L563

Needs more work, so far we just apply any constraints to the inner schema if the validator type is "nullable", we can add more rules.

The good thing is that if the final pydantic-core schema is invalid, pydantic-core will raise an error with details - pydantic-core schema is validated with pydantic-core before creating a validator.

Reply:
Cool! I mean, I wouldn't expect it to just apply to deeply nested types inside, I don't even think that would be good idea :sweat_smile: , it was just curiosity of how it would work. But it's great that pydantic-core will raise an error if something doesn't work or make sense. :rocket:

Reply:
> constraints that don't apply to a list should be automatically forwarded to it's items
Just curious was this part of the original plan? Is the advantage just less boilerplate?

To me this smells like the sort of thing that will cause confusion amongst users and might have edge cases (like the nested application that @tiangolo mentions).

Would you consider not introducing this for now and maybe adding it later @samuelcolvin ?

Reply:
yup, completely open to suggestions, please can you create an issue to discuss it?

Reply:
https://github.com/pydantic/pydantic/issues/4707